File: payola/s2js/runtime/src/main/scala/scala/Array.scala
package s2js.runtime.scala

import s2js.compiler.NativeJs

object Array {
    @NativeJs("""
        var a = new scala.Array(jsArray.length);
        a.internalArray = jsArray;
        return a;
    """)
    def fromNative(jsArray: AnyRef): Array = null

    def apply(xs: Any*): Array = {
        val array = new Array(xs.length)
        var i = 0
        for (x <- xs.iterator) {
            array(i) = x;
            i += 1
        }
        array
    }
}

class Array(val length: Int) {
    @NativeJs("[]")
    var internalArray = null

    @NativeJs("return self.internalArray[i];")
    def apply(i: Int): Any = null

    @NativeJs("return self.internalArray.slice(0);")
    override def clone(): Array = null

    @NativeJs("self.internalArray[i] = x;")
    def update(i: Int, x: Any): Unit = ()
}

File: payola/web/public/javascripts/s2js/runtime/scala/Array.js
goog.provide('scala.Array');
scala.Array = function(length) {
var self = this;
self.length = length;
self.internalArray = [];
};
scala.Array.prototype.$apply = function(i) {
var self = this;
return self.internalArray[i];};
scala.Array.prototype.clone = function() {
var self = this;
return self.internalArray.slice(0);};
scala.Array.prototype.update = function(i, x) {
var self = this;
self.internalArray[i] = x;};
scala.Array.prototype.foreach = function(f) {
var self = this;
self.internalArray.forEach(f);};
scala.Array.prototype.metaClass_ = new s2js.MetaClass('scala.Array', []);
scala.Array.fromNative = function(nativeArray) {
var self = this;

        var a = new scala.Array(nativeArray.length);
        a.internalArray = nativeArray;
        return a;
    };
scala.Array.$apply = function() {
var self = this;
var xs = scala.Array.fromNative([].splice.call(arguments, 0, arguments.length - 0));
return self.fromNative(xs.internalArray);};
scala.Array.metaClass_ = new s2js.MetaClass('scala.Array', []);
